### run_DEXSeq.R ##################################################################################
# This script takes a set of count tables generated by previous steps and run DEXSeq package
# to indentify differentially expressed exons between angiogenesis-induced and control condition
# 

### PREAMBLE ######################################################################################
# Settings for environment

# Define working directory
setwd("/work/angiogenesis/rodrigoguarischi/git/scripts/");

# Add some locally installed libraries to path
.libPaths("~/R/lib/");

# Load required libraries and datasets
library( DEXSeq );
library( org.Mm.eg.db );
library( RColorBrewer );

### FUNCTIONS #####################################################################################

# Load function mergeDataframesByRownames
source("mergeDataframesByRownames.R");

### DATA ANALYSIS #################################################################################

# Initialize object 'tmp' to hold all temporary objects
tmp <- list();

# Initialize empty count data frame to be populated
tmp$counts.df <- data.frame();

# Sum technical replicates and write on files. DEXSeq experts a list of files : (
for ( file in list.files(path="aux_files/counts_by_exon", pattern="*.txt$", full.names = TRUE) ) {

  print( paste0( "Processing file: ", basename(file) ) );
  
  tmp$sample.count <- read.table(
    file = file,
    sep = "\t",
    header = FALSE,
    row.names = 1
    );

  # Assing column name
  colnames(tmp$sample.count) <- gsub(".txt$", "", basename(file) );
  
  # Merge two data frames by rownames  
  tmp$counts.df <- mergeDataframesByRownames(tmp$counts.df, tmp$sample.count);
  };

# Keep row order to write output files
tmp$row.order <- read.table(
  file = file,
  sep = "\t",
  header = FALSE,
  stringsAsFactors = FALSE
  )[, 1];

# Remove variable that will not be used anymore
rm( file );

# Sum counts based in the column names
tmp$allColSamples <- gsub( "-\\d+$", "", colnames(tmp$counts.df) );
tmp$sp <- split( seq( along = tmp$allColSamples ), tmp$allColSamples );
tmp$countdata <- sapply(tmp$sp, function(columns) rowSums( tmp$counts.df[, columns, drop=FALSE] ) );

# Change '_' for '.' on sample R12.5
colnames(tmp$countdata) <- gsub(
  pattern = "R12_5", 
  replacement = "R12.5", 
  x = colnames(tmp$countdata)
  );

# Write summed technical replicated files. DEXSeq experts a list of files : (
dir.create( "aux_files/counts_by_exon/summed_technical_replicates/", showWarnings = FALSE );
for( column in colnames(tmp$countdata) ){
  
  print( paste0("Printing column: ", column) );
  
  write.table(
    x = tmp$countdata[ tmp$row.order, column],
    file = paste0("aux_files/counts_by_exon/summed_technical_replicates/" , column, ".txt"),
    row.names = TRUE,
    col.names = FALSE,
    quote = FALSE,
    sep = "\t"
    );
  
  }

# Remove variable that will not be used anymore
rm( column );

##### Build dxdFull Object #####

# Initialize object 'tmp' to hold all temporary objects
tmp <- list();

tmp$tech.summed.count.files = list.files( path="aux_files/counts_by_exon/summed_technical_replicates", pattern=".txt$", full.names=TRUE);
tmp$sample.order = gsub(".txt$", "", basename(tmp$tech.summed.count.files) );

# Create a factor defining each experimental group (P = Control or R = Retinopathy/Experimetnal)
tmp$condition.factor <- factor(
  gsub(
    pattern = "^(R|P)(.*)_(1|2)$",
    replacement = "\\1", 
    x = tmp$sample.order
    )
  );

# Create a factor defining each experimental time (12, 12.5, 15 or 17)  
tmp$time.factor <- factor(
  gsub(
    pattern = "^(R|P)(.*)_(1|2)$",
    replacement = "\\2",
    x = tmp$sample.order
    )
  );

# Create a factor defining experimental batches (1 or 2)  
tmp$batch.factor <- factor(
  gsub(
    pattern = "^(R|P)(.*)_(1|2)$",
    replacement = "\\3",
    x = tmp$sample.order
    )
  );

# Create data frame coldata with metadata
tmp$coldata <- data.frame(
  condition = tmp$condition.factor,
  time = tmp$time.factor,
  batch = tmp$batch.factor,
  row.names = tmp$sample.order
  );

# Build dxdFull object
dxdFull <- DEXSeqDataSetFromHTSeq(
  countfiles = tmp$tech.summed.count.files,
  sampleData = tmp$coldata,
  design = ~ batch + sample + exon + time:exon + condition:exon,
  flattenedfile = "aux_files/counts_by_exon/ensembl_genes.dexseq.gff"
  );


### Run DEXSeq pipeline and get DE exons for each time

# Initialize object 'de' that will hold data from differential expression
dex <- list();

dex$results <- list();

# Create directory to store output from DEXSeq
dir.create( "../../mRNA/DEXSeqReport/", showWarnings = FALSE );

# Define number of cores to use
tmp$cores <- 40;

for ( current.time in levels(dxdFull$time) ){

  print( paste0( "Calculating DE exons for sample: R", current.time ) );
  
  # Subset dxdFull
  tmp$dxd <- dxdFull[ , dxdFull$condition == 'P' | dxdFull$time == current.time ];
  
  # If testing R12.5 a error stating "not full rank error" will arise. 
  # Due to experimental design we can consider R12.5 as R12 and this will fix this error
  if( current.time == 12.5 ){
    colData(tmp$dxd)$time <- factor(
      gsub(
        pattern = "12.5",
        replacement = "12",
        x = colData(tmp$dxd)$time
        )
      )
    };
  
  # "refactor" the factors, in case that levels have been dropped
  tmp$dxd$condition <- droplevels(tmp$dxd$condition);
  tmp$dxd$time  <- droplevels(tmp$dxd$time);
  
  # Make sure that control is the first level in the treatment factor
  tmp$dxd$condition <- relevel(tmp$dxd$condition, "P");
  tmp$dxd$time <- relevel(tmp$dxd$time, "12");
  
  # Call DEXSeq pipeline
  dex$results[[ paste0("R", current.time) ]] <- DEXSeq(
    object = tmp$dxd,
    fullModel = ~ batch + sample + exon + time:exon + condition:exon,
    reducedModel = ~ batch + sample + exon + time:exon,
    BPPARAM = MulticoreParam( workers = tmp$cores )
    );

  # Subset genes with Differential Exon Usage (DEU)
  tmp$geneIDs <- unique( subset( dex$results[[ paste0("R", current.time) ]], padj < 0.05 & abs(log2fold_R_P) > 1 )$groupID );
  
  # Create array to converte ENSEMBL IDS on Gene Symbols
  tmp$ens2symbol <- mapIds(
    x = org.Mm.eg.db,
    keys = unlist( strsplit( tmp$geneIDs, "\\+" ) ),
    keytype = "ENSEMBL",
    column = "SYMBOL",
    multiVals="first"
    );
  
  # Create data.frame to add to the report
  tmp$extraCols <- data.frame(
    Symbols = sapply(
      tmp$geneIDs,
      function(x) paste( tmp$ens2symbol[ unlist( strsplit( x, "\\+" ) ) ], collapse = "+"),
      USE.NAMES = FALSE
      ),
    row.names = tmp$geneIDs
    );
  
  # Create output dir
  tmp$dexseq_report <- paste0("../../mRNA/DEXSeqReport/R", current.time );
  dir.create(
    path = tmp$dexseq_report,
    showWarnings = FALSE
    );

  # Write output as HTML files
  DEXSeqHTML(
    object = dex$results[[ paste0("R", current.time) ]],
    genes = tmp$geneIDs,
    path = tmp$dexseq_report,
    file = "index.html",
    BPPARAM = MulticoreParam( workers = tmp$cores ),
    color = brewer.pal( n=8, name = "Set2" )[ c(5,4) ],
    extraCols = tmp$extraCols
    );
  
  }

# Get list of DEUs
unique( unlist(lapply( dex$results, function(x) rownames( subset(x, padj < 0.05 & abs(log2fold_R_P) > 1) ) ) ) );

# Get list of genes with DEU
unique( unlist(lapply( dex$results, function(x) subset(x, padj < 0.05 & abs(log2fold_R_P) > 1)$groupID ) ) );
